### PR TITLE
feat(diff): Implement marker-based document diffing engine (fixes #5)

### LINKED ISSUE #5: [Phase 2] Task 2.2: Document Diffing Engine
## Description
Implement core diffing logic to detect additions, deletions, content changes, marker changes, and section movements.

## Dependencies
- Task 2.1: YAML Loader & Validator

## Acceptance Criteria
- Function to compare two Document versions
- Detect added sections (new IDs in new version)
- Detect deleted sections (IDs present only in old version)
- Detect content changes (same ID, different content)
- Detect marker changes (same ID, different marker)
- Detect section movements (same ID, different parent path)
- Handle deeply nested structures correctly
- Provide diff results in structured format
- Support diffing sections at any nesting level

## Tests
- Unit test: Detect added section at root level
- Unit test: Detect added nested section
- Unit test: Detect deleted section
- Unit test: Detect content change
- Unit test: Detect marker change (renumbering)
- Unit test: Detect section movement (different parent)
- Unit test: Detect multiple changes in one diff
- Unit test: Handle empty documents
- Unit test: Handle documents with no changes
- Unit test: Diff deeply nested structures (5+ levels)
- Integration test: Diff two versions of example document
- Acceptance test: Diff matches expected output format

## Files to Create
- `src/yaml_diffs/diff.py`
- `src/yaml_diffs/diff_types.py` (DiffResult, ChangeType enums)
- `tests/test_diff.py`
- `examples/document_v1.yaml`
- `examples/document_v2.yaml` (with various changes)

## Phase
Phase 2: Core Functionality

### PR DESCRIPTION
**Pull request by noamoss**
## Summary

This PR implements a marker-based document diffing engine that compares two Document versions to detect additions, deletions, content changes, movements, and renames. The implementation uses markers (not IDs) as the primary identifiers for matching sections, which aligns with the project's purpose where documents come with markers but IDs are calculated in preprocessing.

## Related Issues

Fixes #5

### Core Diffing Engine
- **New module: diff_types.py** - Diff result types and enums
  - `ChangeType` enum: ADDED, DELETED, CONTENT_CHANGED, MOVED, RENAMED, UNCHANGED
  - `DiffResult` model - Represents individual changes with marker paths and ID paths
  - `DocumentDiff` model - Complete diff results with aggregated counts

- **New module: diff.py** - Core diffing logic
  - `diff_documents()` - Main function to compare two Document versions
  - `_validate_unique_markers()` - Validates no duplicate markers at same nesting level
  - `_build_marker_map()` - Builds marker+path → section mapping (hybrid approach)
  - `_calculate_content_similarity()` - Content similarity scoring for rename detection
  - `_find_moved_sections()` - One-to-one matching to detect moved sections (prevents cartesian product)

### Schema and Model Updates
- **Updated: schema/legal_document_spec.yaml** - Made `marker` required field
- **Updated: models/section.py** - Made `marker` required (not optional)
- **Updated: __init__.py** - Export diff functions and types

### Bug Fixes
- **Bug 1**: Fixed cartesian product issue in `_find_moved_sections` - uses one-to-one matching
- **Bug 2**: Fixed missing RENAMED entries for moved sections with title changes

### Documentation
- **Updated: AGENTS.md** - Document marker-based diffing approach
- **Updated: README.md** - Update key features to reflect marker-based diffing

### Tests
- **New: tests/test_diff.py** - 19 comprehensive tests covering all change types
- **New: tests/test_diff_bugs.py** - 2 tests for bug fixes
- **Updated: tests/test_models.py** - Updated to require markers
- **Updated: tests/test_schema_validation.py** - Updated to reflect marker requirement

### Examples
- **New: examples/document_v1.yaml** - Base document for integration testing
- **New: examples/document_v2.yaml** - Modified version demonstrating all change types
- **Updated: examples/complex_document.yaml** - Added missing marker

## Testing

- [x] All 125 tests pass (including 21 new diff tests)
- [x] Pre-commit hooks pass
- [x] Linting passes (ruff)
- [x] Type checking passes (mypy)
- [x] Integration tests with example documents
- [x] Bug fix verification tests pass

## Acceptance Criteria Met

- [x] Function to compare two Document versions
- [x] Detect added sections (new markers in new version)
- [x] Detect deleted sections (markers present only in old version)
- [x] Detect content changes (same marker+path, different content)
- [x] Detect section movements (same marker, different parent path)
- [x] Detect title changes/renaming (same marker+path+content, different title)
- [x] Handle deeply nested structures correctly (5+ levels)
- [x] Provide diff results in structured format
- [x] Support diffing sections at any nesting level
- [x] Validate duplicate markers at same nesting level (raises error)
- [x] Handle multiple changes per section (e.g., MOVED + CONTENT_CHANGED)
- [x] Use hybrid path approach (marker paths for matching, ID paths for tracking)

## Key Design Decisions

1. **Marker-Based Matching**: Markers are the primary identifiers for diffing (not IDs), as documents come with markers but IDs are calculated in preprocessing.

2. **Hybrid Path Approach**: Uses marker paths for matching sections and ID paths for tracking/reference.

3. **Multiple Change Types**: A single section can have multiple change types (e.g., MOVED + CONTENT_CHANGED as separate entries).

4. **Content Similarity**: Uses 80% similarity threshold for detecting moved sections with content changes.

5. **One-to-One Matching**: Prevents cartesian product issues when multiple sections share the same marker.

## Files Changed

- src/yaml_diffs/diff.py (new, 417 lines)
- src/yaml_diffs/diff_types.py (new, 140 lines)
- src/yaml_diffs/models/section.py (updated)
- src/yaml_diffs/schema/legal_document_spec.yaml (updated)
- src/yaml_diffs/__init__.py (updated)
- tests/test_diff.py (new, 824 lines)
- tests/test_diff_bugs.py (new, 159 lines)
- tests/test_models.py (updated)
- tests/test_schema_validation.py (updated)
- examples/document_v1.yaml (new)
- examples/document_v2.yaml (new)
- examples/complex_document.yaml (updated)
- AGENTS.md (updated)
- README.md (updated)

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/yaml_diffs/diff.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+        if section.sections:
+            new_path = parent_path + (section.marker,)
+            _validate_unique_markers(section.sections, new_path)
+
+
+def _build_marker_map(
+    sections: list[Section],
+    parent_marker_path: tuple[str, ...] = (),
+    parent_id_path: Optional[list[str]] = None,
+) -> dict:
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Type Hint Issue**: The return type `dict` is too generic. Should use a more specific type annotation for better type safety and IDE support.

**Suggested Fix**:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    MarkerMapKey = tuple[str, tuple[str, ...]]
    MarkerMapValue = tuple[Section, tuple[str, ...], list[str]]
    MarkerMap = dict[MarkerMapKey, MarkerMapValue]

def _build_marker_map(
    sections: list[Section],
    parent_marker_path: tuple[str, ...] = (),
    parent_id_path: Optional[list[str]] = None,
) -> MarkerMap:
```

Alternatively, use a type alias at module level for clarity.

File: src/yaml_diffs/diff.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    intersection = words1 & words2
+    union = words1 | words2
+
+    if not union:
+        return 1.0
+
+    return len(intersection) / len(union)
+
+
+def _find_moved_sections(
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Type Hint Issue**: Function parameters and return type are too generic. Should use specific types for better type safety.

**Suggested Fix**:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    MarkerMapKey = tuple[str, tuple[str, ...]]
    MarkerMapValue = tuple[Section, tuple[str, ...], list[str]]
    MarkerMap = dict[MarkerMapKey, MarkerMapValue]

def _find_moved_sections(
    unmatched_old: MarkerMap,
    unmatched_new: MarkerMap,
) -> list[tuple[MarkerMapKey, MarkerMapKey]]:
```

This provides better type checking and makes the function signature clearer.

File: src/yaml_diffs/diff.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+    Examples:
+        >>> old_key = ("1", ("פרק א'",))
+        >>> new_key = ("1", ("פרק ב'",))
+        >>> matches = _find_moved_sections({old_key: ...}, {new_key: ...})
+        >>> assert (old_key, new_key) in matches
+    """
+    matches: list = []
+
+    # Build marker -> keys mapping for new document
+    marker_to_new_keys: dict = {}
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Type Hint Issue**: Variable `marker_to_new_keys` should have a more specific type annotation.

**Suggested Fix**:
```python
marker_to_new_keys: dict[str, list[MarkerMapKey]] = {}
```

This improves type checking and makes the data structure clearer.

**Review by noamoss — COMMENTED** [review]
## Code Review Summary

Great work on implementing the marker-based document diffing engine! The implementation is solid and well-tested. Below is a summary of findings:

### Issues Found

1. **Type Hints (5 issues)**: Several functions and variables use generic `dict` and `list` types instead of more specific type annotations. While the code works correctly, more specific types would improve:
   - Type checking with mypy
   - IDE autocomplete and error detection
   - Code self-documentation

2. **Performance Consideration (1 issue)**: Using `.pop(0)` on a list is O(n). For large documents with many sections sharing the same marker, consider using `collections.deque` for O(1) operations.

3. **Code Clarity (1 issue)**: The logic flow in `diff_documents` uses multiple `if` statements that are functionally correct but could be clearer with `elif` to make the mutually exclusive nature explicit.

### Positive Aspects

✅ **Excellent Test Coverage**: Comprehensive test suite covering all change types, edge cases, and bug fixes. The tests are well-organized and clearly named.

✅ **Good Documentation**: Clear docstrings with examples, helpful comments explaining bug fixes, and good inline documentation.

✅ **Solid Algorithm Design**: The marker-based approach with hybrid path tracking is well-thought-out. The one-to-one matching fix for the cartesian product issue shows good problem-solving.

✅ **Proper Error Handling**: Validation of duplicate markers with clear error messages.

✅ **Hebrew Support**: The implementation correctly handles Hebrew text throughout, which is important for this use case.

✅ **Bug Fixes Well-Documented**: The comments explaining the bug fixes (cartesian product, moved+renamed) are helpful for future maintainers.

### Recommendations

The issues found are minor and mostly relate to code quality improvements rather than bugs. The code is production-ready as-is, but addressing the type hints would be beneficial for long-term maintainability.

**Overall Assessment**: This is a high-quality implementation that meets the requirements. The suggested improvements are optional enhancements that would make the code even better.

**Comment by noamoss** [comment]
## 📋 Review Feedback Addressed

I've applied the following changes based on code review:

### ✅ Changes Made:

1. **Type Hints (5 comments)** - Added specific type annotations using TYPE_CHECKING
   - Created MarkerMap, MarkerMapKey, MarkerMapValue type aliases
   - Updated function signatures: _build_marker_map(), _find_moved_sections()
   - Updated variable types: mapping, matches, marker_to_new_keys
   - Improves type checking with mypy and IDE support
   - (comments #2554776344, #2554776551, #2554776867, #2554777096, #2554777220)

2. **Performance Optimization (1 comment)** - Use collections.deque instead of list.pop(0)
   - Changed from O(n) list.pop(0) to O(1) deque.popleft()
   - Better performance for large documents with many sections sharing the same marker
   - (comment #2554777517)

3. **Code Clarity (1 comment)** - Use elif for mutually exclusive conditions
   - Changed if/if/if to if/elif/else in exact matches section
   - Makes the mutually exclusive nature of change types (CONTENT_CHANGED, RENAMED, UNCHANGED) explicit
   - (comment #2554778164)

4. **Bug Fix (1 comment)** - Always detect content changes for moved sections
   - Fixed issue where CONTENT_CHANGED was only added when similarity >= 0.8 and title unchanged
   - Now always adds CONTENT_CHANGED when content changed, regardless of similarity or title changes
   - Ensures content edits are never lost for moved sections, even when title changes or content is substantially rewritten
   - (comment #2554771461)

### ✅ Validation:
- All 21 tests passed
- Linting passes (ruff)
- Type checking passes (mypy)
- Changes committed and pushed

---
Ready for re-review! 🚀

**Comment by noamoss** [comment]
✅ **Addressed comment #2554771461**

I've verified that the fix for detecting content edits in moved sections is working correctly. The code now always adds `CONTENT_CHANGED` when content changes, regardless of:
- Title changes
- Content similarity score (< 0.8 or >= 0.8)

**Verification:**
- Added 2 new test cases that specifically test the scenarios mentioned in the comment
- Both tests pass, confirming that `modified_count` is correctly incremented
- Content changes are never lost, even when sections move with title changes or substantial rewrites

**Test cases added:**
1. `test_moved_section_with_content_and_title_change` - Verifies MOVED + CONTENT_CHANGED when both content and title change
2. `test_moved_section_with_substantially_rewritten_content` - Verifies CONTENT_CHANGED is detected even with low similarity (< 0.8)

The fix was already applied in the previous commit (78c329c), and these tests confirm it's working as expected.

File: src/yaml_diffs/diff.py [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+                old_id_path=old_id_path,
+                new_id_path=new_id_path,
+            )
+        )
+
+        # BUG FIX: Check for title change (rename) when content is unchanged.
+        # Moved sections with title changes should record both MOVED and RENAMED
+        # to be consistent with non-moved sections. This ensures title change
+        # information is not lost when a section moves.
+        if title_changed and not content_changed:
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Inconsistency Issue**: There's a behavioral inconsistency between exact matches and moved sections regarding RENAMED when both content and title change.

**Current behavior:**
- **Exact matches** (line 266): Records RENAMED even when content also changed (both CONTENT_CHANGED and RENAMED are recorded)
- **Moved sections** (line 325): Only records RENAMED when content is unchanged (if both change, only CONTENT_CHANGED is recorded)

**The comment on line 322 says "to be consistent with non-moved sections"** but the behavior is actually inconsistent.

**Test evidence:**
- `test_diff_exact_match_with_both_content_and_title_change` (line 335) expects both CONTENT_CHANGED and RENAMED
- `test_moved_section_with_content_and_title_change` (line 160) expects only CONTENT_CHANGED (not RENAMED)

**Question**: Is this intentional design, or should moved sections also record RENAMED when both content and title change to be truly consistent?

**If intentional**: The comment on line 322 should be updated to clarify that moved sections have different behavior (RENAMED only for title-only changes).

**If should be consistent**: Update line 325 to remove the `and not content_changed` condition, similar to exact matches.
